waku/v2/discv5/discover_test.go

```diff
@@ -102,3 +152,276 @@ func TestDiscV5(t *testing.T) {
 	require.False(t, peerconn3.HasPeer(host4.ID())) //host4 should not be discoverable, rather filtered out.
 
 }
+
+func TestDiscV5WithCapabilitiesFilter(t *testing.T) {
+
+	// H1
+	host1, _, prvKey1 := tests.CreateHost(t)
+	udpPort1, err := tests.FindFreeUDPPort(t, "127.0.0.1", 3)
+	require.NoError(t, err)
+	ip1, _ := tests.ExtractIP(host1.Addrs()[0])
+	l1, err := tests.NewLocalnode(prvKey1, ip1, udpPort1, wenr.NewWakuEnrBitfield(true, true, true, true), nil, utils.Logger())
+	require.NoError(t, err)
+	peerconn1 := NewTestPeerDiscoverer()
+	d1, err := NewDiscoveryV5(prvKey1, l1, peerconn1, prometheus.DefaultRegisterer, utils.Logger(), WithUDPPort(uint(udpPort1)))
+	require.NoError(t, err)
+	d1.SetHost(host1)
+
+	// H2
+	host2, _, prvKey2 := tests.CreateHost(t)
+	ip2, _ := tests.ExtractIP(host2.Addrs()[0])
+	udpPort2, err := tests.FindFreeUDPPort(t, "127.0.0.1", 3)
+	require.NoError(t, err)
+	l2, err := tests.NewLocalnode(prvKey2, ip2, udpPort2, wenr.NewWakuEnrBitfield(true, true, false, true), nil, utils.Logger())
+	require.NoError(t, err)
+	peerconn2 := NewTestPeerDiscoverer()
+	d2, err := NewDiscoveryV5(prvKey2, l2, peerconn2, prometheus.DefaultRegisterer, utils.Logger(), WithUDPPort(uint(udpPort2)))
+	require.NoError(t, err)
+	d2.SetHost(host2)
+
+	// H3
+	host3, _, prvKey3 := tests.CreateHost(t)
+	ip3, _ := tests.ExtractIP(host3.Addrs()[0])
+	udpPort3, err := tests.FindFreeUDPPort(t, "127.0.0.1", 3)
+	require.NoError(t, err)
+	l3, err := tests.NewLocalnode(prvKey3, ip3, udpPort3, wenr.NewWakuEnrBitfield(true, true, false, true), nil, utils.Logger())
+	require.NoError(t, err)
+	peerconn3 := NewTestPeerDiscoverer()
+	d3, err := NewDiscoveryV5(prvKey3, l3, peerconn3, prometheus.DefaultRegisterer, utils.Logger(), WithUDPPort(uint(udpPort3)))
+	require.NoError(t, err)
+	d3.SetHost(host3)
+
+	defer d1.Stop()
+	defer d2.Stop()
+	defer d3.Stop()
+
+	err = d1.Start(context.Background())
+	require.NoError(t, err)
+
+	err = d2.Start(context.Background())
+	require.NoError(t, err)
+	// Set boot nodes for node2 after the DiscoveryV5 was created
+	err = d2.SetBootnodes([]*enode.Node{d1.localnode.Node()})
+	require.NoError(t, err)
+
+	err = d3.Start(context.Background())
+	require.NoError(t, err)
+	// Set boot nodes for node3 after the DiscoveryV5 was created
+	err = d3.SetBootnodes([]*enode.Node{d2.localnode.Node()})
+	require.NoError(t, err)
+
+	// Desired node capabilities
+	filterBitfield := wenr.NewWakuEnrBitfield(false, false, true, false)
+	iterator3, err := d3.PeerIterator(FilterCapabilities(filterBitfield))
+	require.NoError(t, err)
+	require.NotNil(t, iterator3)
+
+	time.Sleep(2 * time.Second)
+
+	// Check node were discovered by automatic discovery
+	require.True(t, peerconn3.HasPeer(host1.ID()) && peerconn3.HasPeer(host2.ID()))
+
+	peers, err := discoverFilterOnDemand(iterator3, 1)
```

Comment:
I am wondering if we use existing `filterPredicate` option to filter discoveredPeers and let discovery run rather than writing similar method?

Reply:
It would probably work. FilterPredicate, DefaultPredicate were covered. My curiosity was: could I actually stack discovery and filter methods together ? First use "normal" discovery and check, second step use onDemand to filter out and check again. Now I understand filter just goes through already discovered nodes. Right ? Test code runs very fast, so these two step should yield same result with same nodes. If we add another node and let discovery to run longer, we may end up with the second check having different result.  I had brief discussion with @richard-ramos about the iterator being never ending. Perhaps it is on purpose, because we filter from dynamic pool of nodes.

Reply:
> Now I understand filter just goes through already discovered nodes. Right ?

Not already discovered nodes. Filter is run on the nodes as they are being discovered.

Reply:
> Perhaps it is on purpose, because we filter from dynamic pool of nodes.

Yes, as nodes are discovered iterator is run on them..and discovery doesn't stop as long as discv5 is running.

Reply:
> > Perhaps it is on purpose, because we filter from dynamic pool of nodes.
> 
> Yes, as nodes are discovered iterator is run on them..and discovery doesn't stop as long as discv5 is running.

The behavior I saw is round robin across the same nodes which were already discovered. It was like discovery and iterator don't know about each other. For example 3 hosts and count param 6 in discoverFilterOnDemand, output will be  host1,host2,host3,host1,host2,host3. When count param is 2, no problem host1,host2 will get filtered. If n < count, caller needs to know hosts could potentially repeat and dedup on result is needed.

Reply:
That could be happening because the nodes are just sent regularly in discv5. 
IIRC, the discv5 doesn't keep a track of which nodes have been shared with which peer and hence it may end up sending an already shared node which is why you might be seeing this repeated behaviour of same nodes being reported agian and agian.

This is why the user of discv5 needs to store the nodes and ignore if any node is already reported. In our case that is done by using peerStore by the peerManager.